python/perf-kernels/multreduce_matmul_kernel.py

```diff
@@ -0,0 +1,73 @@
+import triton
+import triton.language as tl
+
+
+# Kernel that implements GEMM with explicit multiply-reduce instructions for small block sizes.
+# Based on **tune_gemm** `matmul_kernel` from commit `cf44637` (see `triton-mlir` branch).
+@triton.jit
+def multreduce_matmul_kernel(a_ptr, b_ptr, c_ptr, bias_ptr, M, N, K, stride_am, stride_ak, stride_bk, stride_bn,
+                             stride_cm, stride_cn, stride_bias, BLOCK_SIZE_M: tl.constexpr, BLOCK_SIZE_N: tl.constexpr,
+                             BLOCK_SIZE_K: tl.constexpr, SPLIT_K: tl.constexpr, GROUP_SIZE_M: tl.constexpr,
+                             BIAS: tl.constexpr, EVEN_K: tl.constexpr):
+    pid = tl.program_id(axis=0)
+    pid_z = tl.program_id(1)
+    num_pid_m = tl.cdiv(M, BLOCK_SIZE_M)
+    num_pid_n = tl.cdiv(N, BLOCK_SIZE_N)
+    if GROUP_SIZE_M == 1:
+        pid_m = pid // num_pid_n
+        pid_n = pid % num_pid_n
+    else:
+        num_pid_in_group = GROUP_SIZE_M * num_pid_n
+        group_id = pid // num_pid_in_group
+        first_pid_m = group_id * GROUP_SIZE_M
+        group_size_m = min(num_pid_m - first_pid_m, GROUP_SIZE_M)
+        pid_m = first_pid_m + (pid % group_size_m)
+        pid_n = (pid % num_pid_in_group) // group_size_m
+    if SPLIT_K == 1:
+        offs_k = tl.arange(0, BLOCK_SIZE_K)
+    else:
+        offs_k = pid_z * BLOCK_SIZE_K + tl.arange(0, BLOCK_SIZE_K)
+    offs_am = (pid_m * BLOCK_SIZE_M + tl.arange(0, BLOCK_SIZE_M))
+    offs_bn = (pid_n * BLOCK_SIZE_N + tl.arange(0, BLOCK_SIZE_N))
+    a_ptrs = a_ptr + offs_am[:, None] * stride_am + offs_k[None, :] * stride_ak
+    b_ptrs = b_ptr + offs_k[:, None] * stride_bk + offs_bn[None, :] * stride_bn
+    if BIAS:
+        bias_ptrs = bias_ptr + offs_am * stride_bias
+        bias = tl.load(bias_ptrs, mask=offs_am < M, other=0.0)
+    acc_dtype = tl.float32 if a_ptr.type.element_ty != tl.int8 else tl.int32
+    accumulator = tl.zeros((BLOCK_SIZE_M, BLOCK_SIZE_N), dtype=acc_dtype)
+    for k in range(0, tl.cdiv(K, BLOCK_SIZE_K * SPLIT_K)):
+        if EVEN_K:
+            a = tl.load(a_ptrs)
+            b = tl.load(b_ptrs)
+        else:
+            a = tl.load(a_ptrs, mask=offs_k[None, :] < K - k * BLOCK_SIZE_K, other=0.0)
+            b = tl.load(b_ptrs, mask=offs_k[:, None] < K - k * BLOCK_SIZE_K, other=0.0)
+        # TODO: `min_dot_size` isn't always (16, 16, 16), it's in fact architecture
+        #       dependent and data type dependent. Please check this source file:
+        #       `${TRITON_ROOT}/third_party/amd/backend/compiler.py`
+        #       (look for `min_dot_size` function).
+        #       Q: How can we implement this `if` statement more precisely?
+        #       A: Maybe we can use an approach similar to `EVEN_K`, in which the kernel caller
+        #          inspect all relevant information and provides a Boolean flag for the kernel.
+        if (BLOCK_SIZE_M < 16 or BLOCK_SIZE_N < 16) or BLOCK_SIZE_K < 16:
```

Comment:
Would this be useful for BLOCK_N or K < 16?

Reply:
In terms of performance I really don't know because I didn't test it for `BLOCK_SIZE_N < 16` or `BLOCK_SIZE_K < 16`. At least it'll make the kernel compile correctly because `tl.dot` won't compile for such sizes.

Please check these code snippets and you'll fully understand:
* https://github.com/triton-lang/triton/blob/86f08228102a29b43e998a457981615753f5ff2f/python/triton/language/semantic.py#L1375
* https://github.com/triton-lang/triton/blob/main/third_party/amd/backend/compiler.py#L14